src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertLinqQueryToForEachTests.cs

```diff
@@ -0,0 +1,406 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeActions.ConvertLinq
+{
+    public class ConvertLinqQueryToForEachTests : AbstractCSharpCodeActionTest
+    {
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+          => new CodeAnalysis.CSharp.ConvertLinq.CSharpConvertLinqQueryToLinqMethodProvider();
+
+        [Fact]
+        public async Task NoConversion_MultipleReferences()
+        {
+            var source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    bool M(IEnumerable<int> nums)
+    {
+        var q =  from int n1 in nums 
```

Comment:
since linq to foreach is moving from expression to statement. won't be it better if it just create a new method and replace linq expression to method call? 

like var q = from ... select ..; 
to 
var q = GetX(....);

so for something like q = (from .. select ...).ToList(); 
it becomes
q = (GetX(...)).ToList();

so, it will first do extract method on the linq expression and then rewirte body of the extracted method.

Reply:
and body will be something like this

foreach(int n1 in nums)
+        {
+            foreach(int n2 in nums)
+            {
+                yield return n1;
+            }
+        }

much simpler to do so since the method only contains linq expression in it.

Reply:
otherwise, I think you always need to introduce intermediate collection like list and then pull all items eagarly like

``` csharp
var list = new List<...>();
foreach(int n1 in nums)
        {
            foreach(int n2 in nums)
            {
                list.Add(n1);
            }
        }
````

and replace linq expression with list
var q = (from ... select ...).ToList() => var q = (list).ToList();